src/hotspot/share/gc/shenandoah/shenandoahSimpleBitMap.cpp

```diff
@@ -0,0 +1,244 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "gc/shenandoah/shenandoahSimpleBitMap.hpp"
+
+ShenandoahSimpleBitMap::ShenandoahSimpleBitMap(size_t num_bits) :
+    _num_bits(num_bits),
+    _num_words((num_bits + (BitsPerWord - 1)) / BitsPerWord),
+    _bitmap(NEW_C_HEAP_ARRAY(uintx, _num_words, mtGC))
+{
+  clear_all();
+}
+
+ShenandoahSimpleBitMap::~ShenandoahSimpleBitMap() {
+  if (_bitmap != nullptr) {
+    FREE_C_HEAP_ARRAY(uintx, _bitmap);
+  }
+}
+
+size_t ShenandoahSimpleBitMap::count_leading_ones(idx_t start_idx) const {
+  assert((start_idx >= 0) && (start_idx < _num_bits), "precondition");
+  size_t array_idx = start_idx >> LogBitsPerWord;
+  uintx element_bits = _bitmap[array_idx];
+  uintx bit_number = start_idx & right_n_bits(LogBitsPerWord);
+  uintx omit_mask = right_n_bits(bit_number);
+  uintx mask = ((uintx) 0 - 1) & ~omit_mask;
+  if ((element_bits & mask) == mask) {
+    size_t counted_ones = BitsPerWord - bit_number;
+    return counted_ones + count_leading_ones(start_idx - counted_ones);
+  } else {
+    // Return number of consecutive ones starting with the_bit and including more significant bits.
+    uintx aligned = element_bits >> bit_number;
+    uintx complement = ~aligned;;
+    return count_trailing_zeros<uintx>(complement);
+  }
+}
+
+size_t ShenandoahSimpleBitMap::count_trailing_ones(idx_t last_idx) const {
+  assert((last_idx >= 0) && (last_idx < _num_bits), "precondition");
+  size_t array_idx = last_idx >> LogBitsPerWord;
+  uintx element_bits = _bitmap[array_idx];
+  uintx bit_number = last_idx & right_n_bits(LogBitsPerWord);
+  // All ones from bit 0 to the_bit
+  uintx mask = right_n_bits(bit_number + 1);
+  if ((element_bits & mask) == mask) {
+    size_t counted_ones = bit_number + 1;
+    return counted_ones + count_trailing_ones(last_idx - counted_ones);
+  } else {
+    // Return number of consecutive ones starting with the_bit and including less significant bits
+    uintx aligned = element_bits << (BitsPerWord - (bit_number + 1));
+    uintx complement = ~aligned;
+    return count_leading_zeros<uintx>(complement);
+  }
+}
+
+bool ShenandoahSimpleBitMap::is_forward_consecutive_ones(idx_t start_idx, idx_t count) const {
```

Comment:
Is this a more general version of count_trailing_ones(), i.e. which takes a count (e.g. end-index) instead of implicitely counting until the end of the bitmap? If so, maybe write one to call the other? Also, it looks like here you are using an iterative approach - good!

Reply:
It would be possible to implement is_forward_consecutive_ones(idx start_idx, idx_t count) as count_leading_ones(start_idx) >= count.  However, that would be less efficient than having the independent implementations.  For example, suppose I want to check whether there are 3 consecutive ones, and I call count_leading_ones() and it happens to count 817 consecutive ones.  I didn't really NEED to count all 817 consecutive ones to find out that I have 3 consecutive ones.